docs: reduce readme (#483)
canonical/pebble: README.md

_Take control of your internal daemons!_

**Pebble** is a lightweight Linux service manager that helps you orchestrate a set of local processes as an organised set. It resembles well-known tools such as _supervisord_, _runit_, or _s6_, in that it can easily manage non-system processes independently from the system services. However, it was designed with unique features such as layered configuration and an HTTP API that help with more specific use cases.

Pebble's key features:

- [Layer](https://canonical-pebble.readthedocs-hosted.com/en/latest/reference/layers/)-based configuration
- Service [dependencies](https://canonical-pebble.readthedocs-hosted.com/en/latest/explanation/service-dependencies/)
- Service [logs](https://canonical-pebble.readthedocs-hosted.com/en/latest/reference/cli-commands/logs/) and [log forwarding](https://canonical-pebble.readthedocs-hosted.com/en/latest/reference/log-forwarding/)
- [Health checks](https://canonical-pebble.readthedocs-hosted.com/en/latest/reference/health-checks/)
- [Notices](https://canonical-pebble.readthedocs-hosted.com/en/latest/reference/notices/) (aggregated events)

> - Identities

Reviewer: We can add a link for Identities now. Maybe the how-to guide?

> - Can be used in machines and [containers](https://canonical-pebble.readthedocs-hosted.com/en/latest/reference/pebble-in-containers/)

Reviewer: Per Tony's earlier suggestion, "VMs" or "virtual machines"

```suggestion
- Can be used in virtual machines and [containers](https://canonical-pebble.readthedocs-hosted.com/en/latest/reference/pebble-in-containers/)
```

Author: Technically it can be used in physical machines as well, that is why I used "machines" in the first place. But physical machines already come with systemd, Pebble probably won't be used there, although it can be. Changing to "virtual machines" then.

- Service [dependencies](https://canonical-pebble.readthedocs-hosted.com/en/latest/explanation/service-dependencies/)
- Service [logs](https://canonical-pebble.readthedocs-hosted.com/en/latest/reference/cli-commands/logs/) and [log forwarding](https://canonical-pebble.readthedocs-hosted.com/en/latest/reference/log-forwarding/)
- [Health checks](https://canonical-pebble.readthedocs-hosted.com/en/latest/reference/health-checks/)
- [Notices](https://canonical-pebble.readthedocs-hosted.com/en/latest/reference/notices/) (aggregated events)
- Identities
- Can be used in machines and [containers](https://canonical-pebble.readthedocs-hosted.com/en/latest/reference/pebble-in-containers/)
- [CLI commands](https://canonical-pebble.readthedocs-hosted.com/en/latest/reference/cli-commands/cli-commands/)
- [HTTP API](https://canonical-pebble.readthedocs-hosted.com/en/latest/explanation/api-and-clients/) with a [Go client](https://pkg.go.dev/github.com/canonical/pebble/client) and a [Python client](https://github.com/canonical/operator/blob/main/ops/pebble.py)

## Quick start

Prerequisites:

> - A Linux machine.

Reviewer: This kinda reads funny to have only a single item rerequisite bulleted list. Perhaps remove "Prerequisites:" and "- A linux machine." and just introduce it with something like this:

```
At any Linux shell:
```

```bash

> go install github.com/canonical/pebble/cmd/pebble@latest

Reviewer: We should figure out which installation approach we want to recommend, and use the same one as in the tutorial. On balance it's probably best to use standard Linux tools and not require Go to be installed, so IMO let's go with what's in the tutorial.

Perhaps we can just link to the installation doc to avoid repeating a bunch of stuff. Maybe:

> At any Linux shell, [install Pebble](https://canonical-pebble.readthedocs-hosted.com/en/latest/how-to/install-pebble/#how-to-install-pebble) and then run the following:
>
>     mkdir -p ~/.config/pebble/layers
>     ...

Author: Although using a link reduces duplicated content, it increases the barrier for people who want to give it a "quick start". The user journey would become:

Found the Pebble repo -> read the README -> click a link and jump to another tab -> copy-paste-run three commands -> switch back to the repo tab -> copy-paste the following command.

I think we want to reduce the friction to a minimum so that users can easily start without any hassle.

Even if we copy the three commands from the [how to install pebble](https://canonical-pebble.readthedocs-hosted.com/en/latest/how-to/install-pebble/#how-to-install-pebble) doc, users still need to replace the version tag.

I think `go install` is the simplest way, just one line of command, no need to replace anything, and it's OK to be used in the README because people who check out the repo are most likely developers.

It's also OK to use a different approach in the official doc because first of all, it's a more "official" way of installing a specific version; and second, the doc's audience includes normal users who are not developers.

So, I'd like to keep the `go install` here. Any objections? 

```bash

Reviewer: Good reasoning -- you've convinced me, thanks. Let's keep `go install`.

mkdir -p ~/.config/pebble/layers
export PEBBLE=$HOME/.config/pebble

echo """\
services:
    demo-service:
        override: replace
        command: sleep 1000
        startup: enabled
""" > $PEBBLE/layers/001-demo-service.yaml

pebble run
```

Read more about Pebble's general model [here](https://canonical-pebble.readthedocs-hosted.com/en/latest/explanation/general-model/).

> For a slightly longer introduction, you can also go through the [official tutorial](https://canonical-pebble.readthedocs-hosted.com/en/latest/tutorial/getting-started/)

Reviewer 2: ```suggestion
For a hands-on introduction to Pebble, we recommend going through the [tutorial](https://canonical-pebble.readthedocs-hosted.com/en/latest/tutorial/getting-started/).
```

Reviewer 2: Because "longer" doesn't necessarily mean "better" so hands-on might be more appealing since users know they'll be getting their feet wet.

Thanks.

## Getting help

> If you need support, start with the [documentation](https://canonical-pebble.readthedocs-hosted.com/en/latest/).

Reviewer 2: To get the most out of Pebble, we recommend starting with the [documentation](https://canonical-pebble.readthedocs-hosted.com/en/latest/).

_Take control of your internal daemons!_

> **Pebble** is a lightweight Linux service manager that helps you orchestrate a set of local service processes as an organized set. It resembles well known tools such as _supervisord_, _runit_, or _s6_, in that it can easily manage non-system processes independently from the system services, but it was designed with unique features that help with more specific use cases.

Reviewer 3: I am not 100% on this, and it's in the original text, but I believe that generally it's "well-known" when you're modifying an object and "well known" when you're using it standalone. So here it should be hyphenated.

Reviewer: Agreed. This is already fixed in the docs (https://canonical-pebble.readthedocs-hosted.com/en/latest/), but let's fix it here too.

Let's also copy the other changes we've done in the new docs paragraph, eg "unique features such as layered configuration and an HTTP API that help with more specific use cases".

Pebble's key features:

> - Service management with [layers](https://canonical-pebble.readthedocs-hosted.com/en/latest/reference/layers/)

Reviewer 3: Does the publishing system provide any way for these to be local links here and transformed into full (or rtd-local) ones when it's published? That would be nicer than having full ones here, and also avoid the need to adjust them when there's a nicer domain than `canonical-pebble.readthedocs-hosted.com`.

Reviewer: What do you mean by "the publishing system"? Isn't the main place this lives the github.com repo page?

Reviewer 3: > What do you mean by "the publishing system"? Isn't the main place this lives the github.com repo page?

Ah, I was thinking that this also ended up somewhere on the docs site, probably mixing it up in my mind with the way that PyPI has the ops README.

So irrelevant, sorry.

Reviewer: I'm also asking on `~Documentation` and `~IS` whether we can get something like `pebble.canonical.com`.

Reviewer: It will actually show up on https://pkg.go.dev/github.com/canonical/pebble, but I don't think many people will be reading it there (you'd have to be looking for the [Go client](https://pkg.go.dev/github.com/canonical/pebble@v1.15.0/client)).

- Service [dependencies](https://canonical-pebble.readthedocs-hosted.com/en/latest/explanation/service-dependencies/)
- Service [logs](https://canonical-pebble.readthedocs-hosted.com/en/latest/reference/cli-commands/logs/) and [log forwarding](https://canonical-pebble.readthedocs-hosted.com/en/latest/reference/log-forwarding/)
- [Health checks](https://canonical-pebble.readthedocs-hosted.com/en/latest/reference/health-checks/)
- [Notices](https://canonical-pebble.readthedocs-hosted.com/en/latest/reference/notices/)
- Identities

> - Can be used in Machines and containers

Reviewer 3: ```suggestion
- Can be used in machines and containers
```

More generally: "machines and containers" feels very Juju-y language. For someone unfamiliar with our systems, maybe it's more something like "used on bare metal, VMs, and containers"?

> - [CLI](https://canonical-pebble.readthedocs-hosted.com/en/latest/reference/cli-commands/cli-commands/), [HTTP API and Go client](https://canonical-pebble.readthedocs-hosted.com/en/latest/explanation/api-and-clients/)

Reviewer: What about two bullet points. Maybe:

```
* CLI commands
* [HTTP API] with a [Go client] and a [Python client]
```

## Quick start

Prerequisites:

- A Linux machine.

> - Python 3.x (used to run a basic HTTP server as a sample service managed by Pebble).

Reviewer 3: Is there some other service that's installed in a typical Ubuntu install that could be used? It feels a bit odd to be saying that you need Python, even though there's the "just for the sample service" explanation right afterwards.

Reviewer: Yeah, I wonder if we should just make the service command a "sleep 1000" here.

```bash
git clone https://github.com/canonical/pebble.git

> cd pebble

Reviewer 3: For a quick start that's for users not Pebble devs, probably it would be cleaner to `go install github.com/canonical/pebble/cmd/pebble@latest` or `sudo snap install pebble` and also adjust the command for running pebble, right? Probably the snap would be the best way, so that you don't have to knowingly be doing anything with Go.

Reviewer: Agreed the either of these is better for normal users than cloning. We also present a different way again in the tutorial (download the pre-compiled binaries and unpack). The snap approach looks tantalisingly simple ... but maybe for the README the `go install ...@latest` is best.

mkdir -p ~/PEBBLE/layers

> export PEBBLE=$HOME/PEBBLE

Reviewer 3: I haven't been following the other docs closely, so maybe this has already been decided on, but I don't really like having a suggestion of an all-caps folder in my home directory. Couldn't it be `$HOME/pebble`? Or better, `$HOME/.pebble`? Or, since this is an example, `/tmp/pebble`?

Reviewer: Agreed. I think we should change this here and in the main docs.

Reviewer: Let's finalise `$HOME/pebble` vs `$HOME/.pebble` in daily sync.

Reviewer: Decision was to use `$HOME/.config/pebble` (and update other docs to be consistent).

Author: Per our discussion in the daily, we decide to use `~/.config/pebble` as pebble dir. The tutorial will be changed to the same as well.

echo """\
summary: Simple layer
description: |
    A simple layer.
services:
    http-server:
        override: replace

>         summary: demo http server

Reviewer 3: We could probably leave this line out to help make this more minimal.

Reviewer: Yeah, let's remove the top-level summary and description as well.

        command: python3 -m http.server 8080
        startup: enabled
""" > $PEBBLE/layers/001-http-server.yaml

> go run ./cmd/pebble run

Author: I agree it will be easier for the eyes. But I'm not sure if it's necessary, because the point here is that users can copy the whole block and run it. Thoughts? @benhoyt @tonyandrewmeyer

> go run ./cmd/pebble run

Author: Per our daily sync, we will add empty lines here.

```

> You can also follow the [Getting started with Pebble tutorial](https://canonical-pebble.readthedocs-hosted.com/en/latest/tutorial/getting-started/).

Reviewer: What about "For a slightly longer introduction, you can also go through the [official tutorial](#)."

## Getting help

> If you need support, start with the [documentation](https://canonical-pebble.readthedocs-hosted.com/en/latest/).

Reviewer 3: I think it'd be worth having a link to the MM channel here as well (or maybe a Matrix channel can be added and be the new place to go?).

Reviewer: How about we point people to open GitHub Issues for now? The Mattermost channel is private to Canonical, and I'm not sure about a Matrix channel just yet.

> To learn more about Pebble, read the following sections:

Reviewer: Some of these seems like duplicates of the "key features" list above. Can we combine the features into that list?

And the Hacking/contributing links could just go in the next section.

Author: OK, I will remove this list. I think one item here adds value - the general model link. I will add a sentence about it in the quick start part maybe.

- [General model](https://canonical-pebble.readthedocs-hosted.com/en/latest/explanation/general-model/)
- [Layer configuration examples](https://canonical-pebble.readthedocs-hosted.com/en/latest/reference/layers/)
- [Container usage](https://canonical-pebble.readthedocs-hosted.com/en/latest/reference/pebble-in-containers/)
- [Layer specification](https://canonical-pebble.readthedocs-hosted.com/en/latest/reference/layer-specification/)
- [API and clients](https://canonical-pebble.readthedocs-hosted.com/en/latest/explanation/api-and-clients/)
- [Hacking / Development](#hacking--development)
- [Contributing](#contributing)

> ## Hacking / development

Reviewer: Maybe `Hacking and development`

Reviewer: I also reckon we should remove the "Have fun! and enjoy the rest of the year!" section ... just seems a bit weird to me.
